BroadcastingInMultiplication.py

```diff
@@ -108,8 +126,144 @@ def arrayMultiplication3(arr1, arr2):
 
-"""
-1. There's no problem with the multiplication of an array by a scalar
-2. Not all the arrays can be broadcasted to match each other
-"""
-# TODO: An exception handling is needed to deal with the "unmatchable" scenarios
\ No newline at end of file
+# Function to generate a random matrix of size N
+def randMat(N):
+    """
+    Returns a random square matrix of size N.
+    Parameters:
+    N (integer): The desired size of the array
+    Returns:
+    ndarray: An array of size N
+    """
+    
+    matrix = np.random.rand(N, N) # generates a random matrix of size N with a random value from a uniform distribution (between 0 and 1) for each entry
+    
+    return matrix
+
+# This following section is for analyzing the excution times of the different implementations of Broadcasting for different scenarios
+# We will limit our analysis to dealing only with two dimensoinal arrays (matrices) for consistency and intrepretability purposes (specifically, square matrices)
+
+# First, generating a random number (scalar) and two sets of random matrices for each test for reliability purposes
+randScalar = random.random() # random scalar
+maxMatSize = 100 # maximum matrix size
+randMs1 = [randMat(n+2) for n in range(maxMatSize)] # first set of random matrices
+randMs2 = [randMat(n+2) for n in range(maxMatSize)] # second set of random matrices
+matSizes = [(n + 2) for n in range(maxMatSize)] # an ordered list of the sizes of the generated matrices
+numMat = len(randMs1) # Number of Matrices in each set
+
+# Second, recordinging the excution times for the different implementations of multiplying a matrix by a scalar
+# Default-Broadcasting Times for Scalar multiplication
+dBts = [tm.timeit(lambda: scalarMultiplication1(M, randScalar), number = 1000) for M in randMs1]
+
+# For-loop-Broadcasting Times for Scalar multiplication
+fBts = [tm.timeit(lambda: scalarMultiplication2(M, randScalar), number = 1000) for M in randMs1]
+
+# Einsum-Broadcasting Times for Scalar multiplication
+eBts = [tm.timeit(lambda: scalarMultiplication3(M, randScalar), number = 1000) for M in randMs1]
+
+# Third, recordinging the excution times for the different implementations of multiplying a matrix by a matrix
+# Default-Broadcasting Times for Matrix multiplication
+dBtm = [tm.timeit(lambda: arrayMultiplication1(randMs1[k], randMs2[k]), number = 1000) for k in range(numMat)]
+
+# For-loop-Broadcasting Times for Matrix multiplication
+fBtm = [tm.timeit(lambda: arrayMultiplication2(randMs1[k], randMs2[k]), number = 1000) for k in range(numMat)]
+
+# Einsum-Broadcasting Times for Matrix multiplication
+eBtm = [tm.timeit(lambda: arrayMultiplication3(randMs1[k], randMs2[k]), number = 1000) for k in range(numMat)]
+
+
+# Lastly, preparing for plotting, plotting, and curve-fitting
+# Defining the function we want to fit (it's an exponential function in this case)
+def exp_func(x, a, b):
```

Comment:
The function we want to fit with would be more like A*x**b. We would want to extract the b parameter from our curve fit.

Reply:
Yeah, that's the function defined. The parameters are returned through the plotting function.